src/core/lib/slice/slice_internal.h

```diff
@@ -23,17 +23,168 @@
+namespace grpc_core {
+
+class SliceRefcount {
+ public:
+  enum Type { STATIC, INTERNED, REGULAR };
+  typedef void (*DestroyerFn)(void*);
+
+  SliceRefcount() = default;
+  SliceRefcount(SliceRefcount::Type type, RefCount* ref, bool ensureRefNonZero,
+                DestroyerFn destroyer_fn, void* destroyer_arg,
+                grpc_slice_refcount* sub)
+      : sub_refcount_(sub),
+        ref_type_(type),
+        ref_(ref),
+#ifndef NDEBUG
+        ensureRefNonZero_(ensureRefNonZero),
+#endif
+        dest_fn_(destroyer_fn),
+        destroy_fn_arg_(destroyer_arg) {
+    // Nothing to do here.
+  }
+  SliceRefcount(grpc_slice_refcount* sub, Type type)
+      : sub_refcount_(sub), ref_type_(type) {
+    // Initializer for static.
+  }
+
+  Type getType() const { return ref_type_; }
+
+  int Eq(const grpc_slice& a, const grpc_slice& b);
+
+  uint32_t Hash(const grpc_slice& slice);
+  void Ref() {
+    if (!ref_) return;
+#ifndef NDEBUG
+    if (ensureRefNonZero_) {
+      ref_->RefNonZero();
+      return;
+    }
+#endif
+    ref_->Ref();
+  }
+  void Unref() {
+    if (!ref_) return;
+    if (ref_->Unref()) {
+      dest_fn_(destroy_fn_arg_);
+    }
+  }
+
+  bool Validate() const { return ref_ != nullptr; }
+
+ private:
+  friend struct ::grpc_slice_refcount;
+
+  grpc_slice_refcount* sub_refcount_ = nullptr;
+  Type ref_type_ = REGULAR;
+  RefCount* ref_ = nullptr;
+#ifndef NDEBUG
+  bool ensureRefNonZero_ = false;
+#endif
+  DestroyerFn dest_fn_ = nullptr;
+  void* destroy_fn_arg_ = nullptr;
+};
+
+}  // namespace grpc_core
+
+struct grpc_slice_refcount {
+  grpc_slice_refcount() { impl_.sub_refcount_ = this; }
+  grpc_slice_refcount(grpc_core::SliceRefcount::Type type,
+                      grpc_core::RefCount* ref, bool ensureRefNonZero,
+                      grpc_core::SliceRefcount::DestroyerFn destroyer_fn,
+                      void* destroyer_arg, grpc_slice_refcount* sub)
+      : impl_(type, ref, ensureRefNonZero, destroyer_fn, destroyer_arg, sub) {
+    // Nothing to do here.
+  }
+  grpc_slice_refcount(grpc_slice_refcount* sub,
+                      grpc_core::SliceRefcount::Type type)
+      : impl_(sub, type) {
+    // Nothing to do here.
+  }
+
+  grpc_slice_refcount* sub_refcount() const { return impl_.sub_refcount_; }
+
+  grpc_core::SliceRefcount impl_;
+};
+
+namespace grpc_core {
+
+struct InternedSliceRefcount {
+  static void Destroy(void* arg) {
+    static_cast<InternedSliceRefcount*>(arg)->DestroyInstance();
+  }
+
+  InternedSliceRefcount(size_t length, uint32_t hash,
+                        InternedSliceRefcount* bucket_next)
+      : base(grpc_core::SliceRefcount::Type::INTERNED, &refcnt,
+             /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        sub(grpc_core::SliceRefcount::Type::REGULAR, &refcnt,
+            /*ensureRefNonZero=*/false, Destroy, this, &sub),
+        length(length),
+        hash(hash),
+        bucket_next(bucket_next) {
+    // Nothing else to do here.
```

Comment:
nit: please remove similar comments here and elsewhere.

Reply:
Removed